testing/src/main/java/io/grpc/testing/GrpcServerRule.java

```diff
@@ -0,0 +1,146 @@
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.testing;
+
+import com.google.common.base.Supplier;
+
+import io.grpc.BindableService;
+import io.grpc.ManagedChannel;
+import io.grpc.Server;
+import io.grpc.inprocess.InProcessChannelBuilder;
+import io.grpc.inprocess.InProcessServerBuilder;
+import io.grpc.stub.AbstractStub;
+
+import org.junit.rules.ExternalResource;
+import org.junit.rules.TestRule;
+
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * {@code GrpcServerRule} is a JUnit {@link TestRule} that starts an in-process gRPC service with
+ * the given service implementation. It is particularly useful for mocking out external gRPC-based
+ * services and asserting that the expected requests were made.
+ *
+ * <p>An {@link AbstractStub} can be created against this service by using the
+ * {@link ManagedChannel} provided by {@link GrpcServerRule#channel()}.
+ *
+ * @param <T> the type of the service impl
+ */
+public class GrpcServerRule<T extends BindableService> extends ExternalResource {
+
+  private final Supplier<T> serviceImplSupplier;
+
+  private ManagedChannel channel;
+  private Server server;
+  private String serverName;
+  private T serviceImpl;
+
+  public GrpcServerRule(Supplier<T> serviceImplSupplier) {
+    this.serviceImplSupplier = serviceImplSupplier;
+  }
+
+  /**
+   * After the test has completed, clean up the channel and server.
+   */
+  @Override
+  protected void after() {
+    serverName = null;
+    serviceImpl = null;
+
+    channel.shutdown();
+
+    try {
+      channel.awaitTermination(1, TimeUnit.MINUTES);
+    } catch (InterruptedException e) {
+      throw new RuntimeException(e);
+    } finally {
+      channel = null;
+    }
+
+    server.shutdown();
+
+    try {
+      server.awaitTermination(1, TimeUnit.MINUTES);
+    } catch (InterruptedException e) {
+      throw new RuntimeException(e);
+    } finally {
+      server = null;
+    }
+  }
+
+  /**
+   * Before the test has started, create the server and channel.
+   */
+  @Override
+  protected void before() throws Throwable {
+    serverName = UUID.randomUUID().toString();
+
+    serviceImpl = serviceImplSupplier.get();
+
+    server = InProcessServerBuilder.forName(serverName)
```

Comment:
I'd want to make it easy to use `directExecutor()`, for the client and server. What is the best way for someone to provide a boolean for choosing direct executor? Maybe a `public GrpcServerRule directExecutor()` method that remembers the configuration and then `return this`?